### ISSUE TITLE
`use:*` Directives Use Elements Whose `ownerDocument` Differs from `window.document`

### ISSUE THREAD

**Issue by ITenthusiasm** [solidjs/solid#1936]
### Describe the bug

The element passed to a [`use:*` directive](https://www.solidjs.com/docs/latest/api#use___) has an [`ownerDocument`](https://developer.mozilla.org/en-US/docs/Web/API/Node/ownerDocument) that differs from the application's Document (i.e., `window.document`), even though the element _belongs_ to the application's Document. This prevents directives like the following from working:

```ts
function listenersDirective(element: HTMLElement) {
  element.ownerDocument.addEventListener("click", handleClick);
  // ...

  onCleanup(() => {
    element.ownerDocument.removeEventListener("click", handleClick);
    // ...
  });
}
```

This discrepancy is problematic in certain types of Applications and NPM Packages related to the Web. It's not immediately clear what exactly `element.ownerDocument` is pointing to. It points to a Document, but it definitely doesn't point to `window.document`.

### Your Example Website or App

https://playground.solidjs.com/anonymous/5ddfcd3c-103f-4283-aff6-6720325bfbd5

### Steps to Reproduce the Bug or Issue

1. Create any valid Solid Directive
2. Compare `element.ownerDocument` in the Solid Directive with the global `window.document` and find that the two are not referentially equivalent

The Solid.js Playground Link provided above exemplifies this problem.

### Expected behavior

The `HTMLElement` passed to any given Solid Directive on the Web should have an `ownerDocument` that's equal to the application's Document (i.e., `window.document`).

### Platform

- OS: MacOS
- Browser: Chrome
- Version: 118.0

### Additional context

This bug is not observable after the application has finished rendering. If you open the devtools and select the element that was given a directive, then `element.ownerDocument === window.document` will yield `true` in the `console` even though it yielded `false` during the call to the Solid directive.

**Comment by ITenthusiasm** [comment]
Since the question was raised on Discord, it's also worth pointing out that elements created with `document.createElement` that have not yet been attached to the DOM still have an `ownerDocument` which points to the application's Document.

```js
const div = document.createElement("div");
console.assert(div.ownerDocument === document); // Succeeds
```

[Svelte's Actions](https://svelte.dev/repl/6c54c21dd879493b9e5099d6cc698691?version=4.2.2) are also able to recognize the correct `ownerDocument` of an element.

**Comment by ITenthusiasm** [comment]
Sidenote: Note sure if #1740 is related since I didn't investigate the code. But if it touches `ownerDocument`s, then this bug is probably worth keeping in mind.

**Comment by ryansolid** [comment]
I think this comes down to how elements are created as Solid isn't doing anything directly to mess with this. `cloneNode` vs `createElement` and when owner is applied. Until cloned nodes are attached then they don't have an owner I'm gathering. We do actions and refs at this time so that they can be passed through. So technically they don't have an owning document at this point. Now it is possible that `document.importNode` would straight up solve this as well (although it is slower).

Until we can make a decision I recommend attaching stuff like this in `onMount`.
https://playground.solidjs.com/anonymous/caad4663-af2b-41b8-888e-0035d8f536f3

**Comment by ITenthusiasm** [comment]
From some quick testing in the browser, it seems to me that nodes created with `cloneNode` still have the proper `ownerDocument`. Does it depend on which specific node was cloned?

---

In any case, using `onMount` seems to work fine for my use case. I'm really thankful that your framework's primitives are usable just about anywhere (compared to something like `React.useEffect`). It makes life a lot easier in situations like these.

If this doesn't get fixed/changed (depending on if it's seen as a bug), it might be worth pointing this out in the documentation. Not sure if it's too obscure/specific to mention, though. 🤷🏾‍♂️

**Comment by ryansolid** [comment]
Yeah I think it is because the elements are coming from template elements. I don't know what else would do it. We really aren't doing anything special.

**Comment by ITenthusiasm** [comment]
Yeah that seems to check out.

In the browser: 
```js
let template = document.createElement("template");
template.innerHTML = "<span>Hello</span>";
console.log(template.content.children[0].ownerDocument === document); // `false`
```

**Comment by ryansolid** [comment]
Well those Svelte Devs may be in for a surprise with Svelte 5 depending on timing of when they fire actions. They might fire them after connected. But Svelte 5 basically adopts Solid's render model.
